README.md
## Bottom Bar With Indicator
A Bottom navigation panel with top indicator.

Click [**here**](https://github.com/tokyolem/bottom_bar_with_indicator/blob/main/example/lib/main.dart) to view the full example.

## Installing
Add it to your `pubspec.yaml` file:
```yml
dependencies:
  bottom_bar_with_indicator: ^0.0.2
```
Install packages from the command line
```
flutter pub get
```
## Usage

You can use this in the following way:

```dart
@override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Custom bottom bar',
      home: Scaffold(
        bottomNavigationBar: BottomBarWithIndicator(
          selectedIndex: _selectedIndex,
          onIndexChanged: (index) {
            setState(() => _selectedIndex = index);
          },
          themeData: BarWithIndicatorThemeData(
            backgroundColor: const Color(0xFFF6F6F6),
            activeColor: const Color(0xFF3C5FE6),
            inactiveColor: const Color(0xFFB1B8C2),
            floating: true,
          ),
          items: const <BottomBarWithIndicatorItem>[
            BottomBarWithIndicatorItem(icon: Icons.delete, label: 'Item 1'),
            BottomBarWithIndicatorItem(icon: Icons.delete, label: 'Item 2'),
            BottomBarWithIndicatorItem(icon: Icons.delete, label: 'Item 3'),
          ],
        ),
      ),
    );
  }
```

### Result:

![ezgif com-video-to-gif](https://github.com/tokyolem/bottom_bar_with_indicator/assets/93796040/437c9680-67da-4251-a57a-377095d2f213)


## Customization

You can customize the widget to your liking. There is a helper `BarWithIndicatorThemeData` class for this.

| Parameter        | Description                                                                                                   |
|------------------|---------------------------------------------------------------------------------------------------------------|
| backgroundColor  | The background color of the navigation bar.                                    |
| activeColor   | The color of the selected item in the navigation bar.                                  |
| inactiveColor      | The color of inactive items in the navigation bar.       |
| floating     | Determines whether the navigation bar has a floating appearance. When set to `true`, the navigation bar will have padding from the parent widget's borders and slightly rounded corners. When set to `false`, the default navigation bar model will be used.                                             |
| cornerRadius          | The corner radius of the navigation bar.                                                  |
| barMargin | The padding of the navigation bar from the edges of the parent widget.      |


CHANGELOG.md
## 0.0.2

* Fix bug with top indicator padding in Desktop apps.
* Improved tab scrolling for mobile devices in landscape mode.
* Minimum version of dart 3.0.0

## 0.0.1

* Initial release.


lib/bottom_bar/bar_constants.dart
import 'dart:ui';

abstract final class BarDimensions {
  /// Value == 0.
  static const double none = 0;

  /// Value == 2.
  static const double extraSmall = 2;

  /// Value == 4.
  static const double small = 4;

  /// Value == 8.
  static const double medium = 8;

  /// Value == 16.
  static const double large = 16;

  /// Value == 32.
  static const double extraLarge = 32;

  /// Value == 90.
  static const double customBarHeight = 90;

  /// Value == 80.
  static const double barItemWidth = 80;

  /// Value == 3.
  static const double barIndicatorStrokeWidth = 3;

  /// Value == 300.
  static const double absoluteCircle = 300;
}

abstract final class BarColors {
  static const Color barColor = Color(0xFFF6F6F6);
  static const Color activeColor = Color(0xFF3C5FE6);
  static const Color inactiveColor = Color(0xFFB1B8C2);
  static const Color labelColor = Color(0xFFB1B8C2);
}

abstract final class BarDurations {

  /// Value == 200 milliseconds.
  static const Duration small =
  Duration(milliseconds: 200);
}


lib/bottom_bar/bar_with_indicator_theme.dart
import 'package:bottom_bar_with_indicator/bottom_bar/bar_with_indicator_theme_data.dart';
import 'package:flutter/material.dart';

/// A widget that defines the theme data for a custom bar.
///
/// The `CustomBarTheme` widget is used to provide the theme data to the
/// descendant widgets in the widget tree. It wraps its `child` widget and
/// allows accessing the `CustomBarThemeData` using `CustomBarTheme.of(context)`.
///
/// The `CustomBarTheme` widget should be placed above the widgets that need
/// to access the custom bar theme data in the widget tree.
class BarWithIndicatorTheme extends StatelessWidget {
  /// Creates a `CustomBarTheme`.
  ///
  /// The `child` and `data` arguments must not be null.
  const BarWithIndicatorTheme({
    required this.child,
    required this.data,
    super.key,
  });

  /// The widget below this `CustomBarTheme` widget in the widget tree.
  final Widget child;

  /// The theme data for the custom bar.
  final BarWithIndicatorThemeData data;

  /// Retrieves the `CustomBarThemeData` from the nearest ancestor
  /// `CustomBarTheme` widget.
  ///
  /// Returns the `CustomBarThemeData` if it exists, otherwise returns
  /// a default `CustomBarThemeData`.
  ///
  /// Typical usage is as follows:
  ///
  /// ```dart
  /// CustomBarThemeData theme = CustomBarTheme.of(context);
  /// ```
  static BarWithIndicatorThemeData of(BuildContext context) {
    final inheritedTheme =
    context.dependOnInheritedWidgetOfExactType<_InheritedCustomBarTheme>();

    final theme = inheritedTheme?.theme.data;
    return theme ?? BarWithIndicatorThemeData();
  }

  @override
  Widget build(BuildContext context) {
    return _InheritedCustomBarTheme(
      theme: this,
      child: child,
    );
  }
}

/// An inherited widget that holds the `CustomBarTheme` data.
class _InheritedCustomBarTheme extends InheritedTheme {
  const _InheritedCustomBarTheme({
    required this.theme,
    required super.child,
  });

  /// The `CustomBarTheme` associated with this inherited widget.
  final BarWithIndicatorTheme theme;

  @override
  Widget wrap(BuildContext context, Widget child) {
    return BarWithIndicatorTheme(data: theme.data, child: child);
  }

  @override
  bool updateShouldNotify(_InheritedCustomBarTheme old) =>
      theme.data != old.theme.data;
}


lib/bottom_bar/bottom_bar_with_indicator.dart
import 'dart:io';

import 'package:bottom_bar_with_indicator/bottom_bar/bar_constants.dart';
import 'package:bottom_bar_with_indicator/bottom_bar/bar_scroll_behavior.dart';
import 'package:bottom_bar_with_indicator/bottom_bar/bar_with_indicator_theme.dart';
import 'package:bottom_bar_with_indicator/bottom_bar/bar_with_indicator_theme_data.dart';
import 'package:flutter/material.dart';

class BottomBarWithIndicator extends StatelessWidget {
  final List<BottomBarWithIndicatorItem> items;
  final int selectedIndex;
  final ValueChanged<int>? onIndexChanged;
  final BarWithIndicatorThemeData? themeData;

  const BottomBarWithIndicator({
    required this.items,
    required this.selectedIndex,
    this.onIndexChanged,
    this.themeData,
    super.key,
  });

  @override
  Widget build(BuildContext context) {
    return BarWithIndicatorTheme(
      data: themeData ?? BarWithIndicatorThemeData(),
      child: Builder(
        builder: (context) {
          return _BarContainer(
            items: items,
            selectedIndex: selectedIndex,
            onIndexChanged: onIndexChanged,
          );
        },
      ),
    );
  }
}

class _BarContainer extends StatefulWidget {
  final List<BottomBarWithIndicatorItem> items;
  final int selectedIndex;
  final ValueChanged<int>? onIndexChanged;

  const _BarContainer({
    required this.items,
    required this.selectedIndex,
    required this.onIndexChanged,
  });

  @override
  State<_BarContainer> createState() => _BarContainerState();
}

class _BarContainerState extends State<_BarContainer>
    with SingleTickerProviderStateMixin {
  late final TabController _tabController;

  @override
  void initState() {
    super.initState();

    _tabController = TabController(length: widget.items.length, vsync: this)
      ..index = widget.selectedIndex;
  }

  @override
  void dispose() {
    _tabController.dispose();

    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    final floating = BarWithIndicatorTheme.of(context).floating;

    return Container(
      width: MediaQuery.of(context).size.width,
      height: BarDimensions.customBarHeight,
      margin: floating
          ? const EdgeInsets.all(BarDimensions.large)
          : BarWithIndicatorTheme.of(context).barMargin,
      decoration: BoxDecoration(
        color: BarWithIndicatorTheme.of(context).backgroundColor,
        borderRadius: floating
            ? const BorderRadius.all(Radius.circular(BarDimensions.large))
            : BarWithIndicatorTheme.of(context).cornerRadius,
      ),
      child: LayoutBuilder(
        builder: (context, constraints) {
          final isScrollable = _isScrollable(constraints);

          return ScrollConfiguration(
            behavior: BarScrollBehavior(),
            child: TabBar(
              isScrollable: isScrollable,
              controller: _tabController,
              splashFactory: NoSplash.splashFactory,
              overlayColor: MaterialStateProperty.resolveWith<Color?>(
                (state) => state.contains(MaterialState.focused)
                    ? null
                    : Colors.transparent,
              ),
              indicator: UnderlineTabIndicator(
                borderSide: BorderSide(
                  color: BarWithIndicatorTheme.of(context).activeColor,
                  width: BarDimensions.barIndicatorStrokeWidth,
                ),
                insets: const EdgeInsets.only(
                  bottom: BarDimensions.customBarHeight -
                      BarDimensions.barIndicatorStrokeWidth -
                      1,
                ),
                borderRadius: const BorderRadius.all(
                  Radius.circular(BarDimensions.extraLarge),
                ),
              ),
              onTap: widget.onIndexChanged,
              tabs: [
                for (var index = 0; index < widget.items.length; index++)
                  _BarItem(
                    item: widget.items[index],
                    itemColor: index == widget.selectedIndex
                        ? BarWithIndicatorTheme.of(context).activeColor
                        : BarWithIndicatorTheme.of(context).inactiveColor,
                    selectedIndex: widget.selectedIndex,
                    itemIndex: index,
                  ),
              ],
            ),
          );
        },
      ),
    );
  }

  bool _isScrollable(BoxConstraints constraints) {
    final isMobile = Platform.isAndroid || Platform.isIOS;

    return isMobile
        ? widget.items.length >= 5 && constraints.maxWidth < 400 ||
            widget.items.length >= 8
        : constraints.maxWidth <
            widget.items.length * BarDimensions.barItemWidth + 10;
  }
}

class _BarItem extends StatelessWidget {
  final BottomBarWithIndicatorItem item;
  final Color itemColor;
  final int selectedIndex;
  final int itemIndex;

  const _BarItem({
    required this.item,
    required this.itemColor,
    required this.itemIndex,
    required this.selectedIndex,
  });

  @override
  Widget build(BuildContext context) {
    return GestureDetector(
      child: AnimatedContainer(
        width: BarDimensions.barItemWidth,
        duration: BarDurations.small,
        child: Column(
          mainAxisSize: MainAxisSize.min,
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            Padding(
              padding: const EdgeInsets.all(BarDimensions.large),
              child: Icon(item.icon, color: itemColor),
            ),
            AnimatedSize(
              duration: BarDurations.small,
              curve: Curves.easeInOut,
              child: SizedBox(
                height: selectedIndex == itemIndex ? null : 0,
                child: Text(
                  item.label,
                  maxLines: 1,
                  overflow: TextOverflow.ellipsis,
                  style: TextStyle(
                    color: BarWithIndicatorTheme.of(context).inactiveColor,
                    fontWeight: FontWeight.bold,
                  ),
                ),
              ),
            ),
          ],
        ),
      ),
    );
  }
}

class BottomBarWithIndicatorItem {
  final IconData icon;
  final String label;

  const BottomBarWithIndicatorItem({
    required this.icon,
    required this.label,
  });
}


lib/bottom_bar/bar_with_indicator_theme_data.dart
import 'package:bottom_bar_with_indicator/bottom_bar/bar_constants.dart';
import 'package:flutter/material.dart';

/// Defines the theme data for the custom navigation bar.
///
/// The `CustomBarThemeData` class allows convenient customization of the
/// navigation bar widget by specifying options such as
/// [backgroundColor], [activeColor], [inactiveColor],
/// [floating], [cornerRadius], [barMargin],
@immutable
class BarWithIndicatorThemeData {
  /// Creates a `CustomBarThemeData` instance.
  ///
  /// The [backgroundColor], [activeColor], [inactiveColor], [floating],
  /// [cornerRadius], and [barMargin] arguments can be used to customize
  /// the appearance and behavior of the navigation bar.
  factory BarWithIndicatorThemeData({
    Color? backgroundColor,
    Color? activeColor,
    Color? inactiveColor,
    bool? floating,
    BorderRadius? cornerRadius,
    EdgeInsets? barMargin,
  }) {
    backgroundColor ??= BarColors.barColor;
    activeColor ??= BarColors.activeColor;
    inactiveColor ??= BarColors.inactiveColor;
    floating ??= false;
    cornerRadius ??= const BorderRadius.all(Radius.zero);
    barMargin ??= EdgeInsets.zero;

    return BarWithIndicatorThemeData._defaultTheme(
      backgroundColor: backgroundColor,
      activeColor: activeColor,
      inactiveColor: inactiveColor,
      floating: floating,
      cornerRadius: cornerRadius,
      barMargin: barMargin,
    );
  }

  /// The background color of the navigation bar.
  final Color backgroundColor;

  /// The color of the selected item in the navigation bar.
  final Color activeColor;

  /// The color of inactive items in the navigation bar.
  final Color inactiveColor;

  /// Determines whether the navigation bar has a floating appearance.
  ///
  /// When set to `true`, the navigation bar will have padding from the parent
  /// widget's borders and slightly rounded corners. When set to `false`, the
  /// default navigation bar model will be used.
  final bool floating;

  /// The corner radius of the navigation bar.
  final BorderRadius cornerRadius;

  /// The padding of the navigation bar from the edges of the parent widget.
  final EdgeInsets barMargin;

  const BarWithIndicatorThemeData._defaultTheme({
    required this.backgroundColor,
    required this.activeColor,
    required this.inactiveColor,
    required this.floating,
    required this.cornerRadius,
    required this.barMargin,
  });
}


example/lib/main.dart
import 'package:bottom_bar_with_indicator/bottom_bar_with_indicator.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(const App());
}

class App extends StatefulWidget {
  const App({super.key});

  @override
  State<App> createState() => _AppState();
}

class _AppState extends State<App> {
  var _selectedIndex = 0;

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Custom bottom bar',
      home: Scaffold(
        bottomNavigationBar: BottomBarWithIndicator(
          selectedIndex: _selectedIndex,
          onIndexChanged: (index) {
            setState(() => _selectedIndex = index);
          },
          themeData: BarWithIndicatorThemeData(
            backgroundColor: const Color(0xFFF6F6F6),
            activeColor: const Color(0xFF3C5FE6),
            inactiveColor: const Color(0xFFB1B8C2),
            floating: true,
          ),
          items: const <BottomBarWithIndicatorItem>[
            BottomBarWithIndicatorItem(icon: Icons.delete, label: 'Item 1'),
            BottomBarWithIndicatorItem(icon: Icons.delete, label: 'Item 2'),
            BottomBarWithIndicatorItem(icon: Icons.delete, label: 'Item 3'),
          ],
        ),
      ),
    );
  }
}
